spinalcordtoolbox/labels.py

```diff
@@ -524,57 +525,28 @@ def project_centerline(img: Image, ref: Image) -> Image:
     :param ref: reference labels
     :returns: image with the new projected labels on the centerline
     """
+    # The centerline of the segmentation, with one point for each S-I slice.
+    centerline_phys = get_centerline(img, space='phys')[1].T  # shape (N, 3), float
 
-    # Checking orientation
-    og_img_orientation = img.orientation
-    if img.orientation != "RPI":
-        img.change_orientation("RPI")
-
-    og_ref_orientation = ref.orientation
-    if ref.orientation != "RPI":
-        ref.change_orientation("RPI")
+    # The label coordinates, in the same physical space.
+    labels_pix = np.argwhere(ref.data)  # shape (L, 3), int
+    labels_phys = ref.transfo_pix2phys(labels_pix, mode='absolute')  # shape (L, 3), float
 
-    # Checking input dimensions
-    if img.data.shape != ref.data.shape:
-        raise ShapeMismatchError(
-            "Input image and referenced labels should have the same dimension",
-            img=img.data.shape,
-            ref=ref.data.shape)
+    # For each label voxel, the centerline index which is closest in physical space.
+    _, indices = KDTree(centerline_phys).query(labels_phys)  # shape (L,), int
 
-    # Fetch resolution
-    nx, ny, nz, nt, px, py, pz, pt = img.dim
-
-    # Extract centerline from segmentation
-    _, arr_ctl, _, _ = get_centerline(img)
-    centerline = arr_ctl.T
-    centerline_real = centerline*np.array([px, py, pz])
+    # Check for collisions
+    if len(set(indices)) != len(indices):
+        logger.warning("Two labels were projected on the same coordinate")
 
-    # Extract referenced coordinates
-    coordinates_ref = ref.getNonZeroCoordinates(sorting='value')
+    # Go from centerline indices back to voxel coordinates in the segmentation image.
+    projected_phys = centerline_phys[indices]  # shape (L, 3), float
+    projected_pix = img.transfo_phys2pix(projected_phys)  # shape (L, 3), int
+    projected_pix = projected_pix.clip(0, np.array(img.data.shape) - 1)
 
-    # Create the output image
-    out = zeros_like(ref)
-
-    # Compute the shortest distance for each referenced points on the centerline
-    for x, y, z, value in coordinates_ref:
-        x_real, y_real, z_real = x * px, y * py, z * pz
-        projection_real = project_point_on_line(point=np.array([x_real, y_real, z_real]), line=centerline_real)
```

Comment:
Presumably once this same logic is applied to `label_regions_from_reference`, there will be no need for the old `project_point_on_line` function?

Reply:
Removed in b44c53c4732417c90f95d2b91d2e590725b6b9ea.